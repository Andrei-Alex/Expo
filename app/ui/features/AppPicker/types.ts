import { icons } from '../../atoms/Icon';

export interface IAppPicker {
  icon?: icons;
  placeholder?: string;
}
