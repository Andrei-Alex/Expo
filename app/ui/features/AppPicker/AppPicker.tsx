import { AppText, Icon } from '../../atoms';
import * as React from 'react';
import { useState } from 'react';
import {
  Button,
  FlatList,
  Modal,
  TouchableWithoutFeedback,
  View
} from 'react-native';
import { colors } from '../../../config';
import { IAppPicker, styles } from './index';
import { MainScreen } from '../../layouts';
import { PickerItem } from '../../components';

/**
 * AppPicker Component
 *
 * A custom picker component for React Native that displays a modal list of selectable items. The component integrates
 * an optional leading icon, and displays the selected item's label or a placeholder when no item is selected. Tapping
 * on the component opens a modal with a list of items, each represented by a `PickerItem`. The user can select an item
 * from the list, which then updates the displayed label and invokes the `onSelectItem` callback with the selected item.
 *
 * ### Usage
 *
 * ```jsx
 * <AppPicker
 *   items={[{ label: 'Item 1', value: 1 }, { label: 'Item 2', value: 2 }]}
 *   selectedItem={{ label: 'Item 1', value: 1 }}
 *   onSelectItem={(item) => console.log(item)}
 *   icon="apps"
 *   placeholder="Select an item"
 * />
 * ```
 *
 * @component
 * @param {IAppPicker} props - The properties passed to the component.
 * @param {(item: item) => void} props.onSelectItem - Callback function that is invoked with the selected item when an item is selected.
 * @param {item} props.selectedItem - The currently selected item, used to display the label in the picker field. If no item is selected, the `placeholder` is displayed instead.
 * @param {item[]} props.items - An array of items that the user can select from. Each item must have a `label` for display and a `value` as a unique identifier.
 * @param {icons} [props.icon] - Optional. The name of the icon to display on the left side of the picker field, from the predefined set of icons.
 * @param {string} [props.placeholder] - Optional. The text to display when no item is selected.
 * @returns {React.ReactElement} A customizable picker component with an optional icon and a modal selection list.
 */
const AppPicker: React.FC<IAppPicker> = ({
  onSelectItem,
  selectedItem,
  items,
  icon,
  placeholder,
  modalTestId = 'modalPickerID',
  pickerTestId = 'pickerTestID',
  style,
  width = '100%'
}) => {
  const [modalVisible, setModalVisible] = useState(false);

  return (
    <>
      <TouchableWithoutFeedback
        onPress={() => setModalVisible(true)}
        testID={pickerTestId}
      >
        <View style={[styles.container, style, [{ width: width }]]}>
          {icon && (
            <Icon
              name={icon}
              size={30}
              iconColor={colors.medium}
              backgroundColor={'transparent'}
              style={styles.icon}
            />
          )}
          <AppText style={selectedItem ? styles.text : styles.placeholder}>
            {selectedItem ? selectedItem.label : placeholder}
          </AppText>
          <Icon
            name={'chevron-down'}
            size={30}
            iconColor={colors.medium}
            backgroundColor={'transparent'}
            style={styles.icon}
          />
        </View>
      </TouchableWithoutFeedback>
      <Modal
        visible={modalVisible}
        animationType={'slide'}
        testID={modalTestId}
      >
        <MainScreen>
          <Button title={'Close'} onPress={() => setModalVisible(false)} />
          <FlatList
            data={items}
            keyExtractor={(item) => item.value.toString()}
            renderItem={({ item }) => (
              <PickerItem
                label={item.label}
                onPress={() => {
                  setModalVisible(false);
                  onSelectItem(item);
                }}
              />
            )}
          />
        </MainScreen>
      </Modal>
    </>
  );
};

export default AppPicker;
