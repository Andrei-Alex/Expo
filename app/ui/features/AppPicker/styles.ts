import { StyleSheet } from 'react-native';

import { colors, fonts } from '../../../config';

export const styles = StyleSheet.create({
  container: {
    backgroundColor: colors.light,
    borderRadius: 25,
    flexDirection: 'row',
    width: '100%',
    padding: 15,
    marginVertical: 10
  },
  icon: {
    marginRight: 10
  },
  text: {
    fontFamily: fonts.primary,
    color: colors.dark,
    fontSize: 18,
    marginRight: 10,
    flex: 1
  }
});
