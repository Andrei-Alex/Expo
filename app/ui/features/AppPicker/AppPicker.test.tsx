import React from 'react';
import renderer from 'react-test-renderer';
import AppPicker from './AppPicker';
import { render, fireEvent } from '@testing-library/react-native';

describe('AppPicker Component', () => {
  const onSelectItemMock = jest.fn();

  it('renders correctly with a placeholder', () => {
    const component = renderer.create(
      <AppPicker
        items={[
          { label: 'Item 1', value: 1 },
          { label: 'Item 2', value: 2 }
        ]}
        placeholder="Select an item"
      />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('opens modal on touchable press', () => {
    const { getByTestId } = render(
      <AppPicker
        items={[
          { label: 'Item 1', value: 1 },
          { label: 'Item 2', value: 2 }
        ]}
        placeholder="Select an item"
      />
    );

    const touchable = getByTestId('pickerTestID');
    fireEvent.press(touchable);

    const modal = getByTestId('modalPickerID');
    expect(modal).toBeTruthy();
  });

  it('selects an item and calls onSelectItem callback', () => {
    const { getByTestId } = render(
      <AppPicker
        items={[
          { label: 'Item 1', value: 1 },
          { label: 'Item 2', value: 2 }
        ]}
        onSelectItem={onSelectItemMock}
        placeholder="Select an item"
      />
    );

    fireEvent.press(getByTestId('pickerTestID'));
    fireEvent.press(getByTestId('Item 1'));

    expect(onSelectItemMock).toHaveBeenCalledWith({
      label: 'Item 1',
      value: 1
    });
  });

  it('closes modal after selecting an item', () => {
    const { getByTestId, queryByTestId } = render(
      <AppPicker
        onSelectItem={onSelectItemMock}
        items={[
          { label: 'Item 1', value: 1 },
          { label: 'Item 2', value: 2 }
        ]}
        placeholder="Select an item"
      />
    );

    fireEvent.press(getByTestId('pickerTestID'));
    fireEvent.press(getByTestId('Item 1'));

    expect(queryByTestId('modal')).toBeNull();
  });
});
