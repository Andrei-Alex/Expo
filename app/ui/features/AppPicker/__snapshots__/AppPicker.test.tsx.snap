// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPicker Component renders correctly with a placeholder 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#f8f4f4",
        "borderRadius": 25,
        "flexDirection": "row",
        "marginVertical": 10,
        "padding": 15,
        "width": "100%",
      },
      undefined,
    ]
  }
  testID="pickerTestID"
>
  <Text
    style={
      [
        {
          "fontFamily": "Avenir",
          "fontSize": 20,
        },
        {
          "color": "#6e6969",
          "flex": 1,
          "fontFamily": "Avenir",
          "fontSize": 18,
          "marginRight": 10,
        },
      ]
    }
  >
    Select an item
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "justifyContent": "center",
        },
        {
          "backgroundColor": "transparent",
          "borderRadius": 15,
          "height": 30,
          "width": 30,
        },
        {
          "marginRight": 10,
        },
      ]
    }
    testID="icon-container-chevron-down"
  >
    <Text />
  </View>
</View>
`;
